oddt/shape.py

```diff
@@ -0,0 +1,249 @@
+
+        ftf : numpy array or None (default = None)
+            Coordinates of the farthest atom
+            to the farthest atom to the molecular centroid
+            If 'None', the point is calculated
+
+        atoms_type : str or None (default None)
+            Type of atoms to be selected from atom_dict
+            If 'None', all atoms are used to calculate shape descriptor
+
+        Returns
+        -------
+        shape_descriptor : numpy array, shape = (12)
+            Array describing shape of molecule
+    """
+    if atoms_type is None:
+        atoms = molecule.atom_dict['coords']
+    else:
+        atoms = molecule.atom_dict[molecule.atom_dict[atoms_type]]['coords']
+
+    if ctd is None:
+        ctd = atoms.mean(0)
+    distances_ctd = norm(atoms - ctd, axis=1)
+
+    if cst is None:
+        cst = atoms[distances_ctd.argmin()]
+    distances_cst = norm(atoms - cst, axis=1)
+
+    if fct is None:
+        fct = atoms[distances_ctd.argmax()]
+    distances_fct = norm(atoms - fct, axis=1)
+
+    if ftf is None:
+        ftf = atoms[distances_fct.argmax()]
+    distances_ftf = norm(atoms - ftf, axis=1)
+
+    distances_list = [distances_ctd, distances_cst, distances_fct, distances_ftf]
+
+    shape_descriptor = np.zeros(12)
+
+    for i, distances in enumerate(distances_list):
+        shape_descriptor[i*3 + 0] = np.mean(distances)
+        shape_descriptor[i*3 + 1] = np.var(distances)
+        shape_descriptor[i*3 + 2] = stats.moment(distances, moment=3)
+
+    return shape_descriptor, (ctd, cst, fct, ftf)
+
+
+def usr(molecule):
+    """Computes USR shape descriptor based on
+        Ballester PJ, Richards WG (2007). Ultrafast shape recognition to search
+        compound databases for similar molecular shapes. Journal of
+        computational chemistry, 28(10):1711-23.
+        http://dx.doi.org/10.1002/jcc.20681
+
+        Parameters
+        ----------
+        molecule : oddt.toolkit.Molecule
+            Molecule to compute USR shape descriptor
+
+        Returns
+        -------
+        shape_descriptor : numpy array, shape = (12)
+            Array describing shape of molecule
+    """
+    return common_usr(molecule)[0]
+
+
+def usr_cat(molecule):
+    """Computes USRCAT shape descriptor based on
+        Adrian M Schreyer, Tom Blundell (2012). USRCAT: real-time ultrafast
+        shape recognition with pharmacophoric constraints. Journal of
+        Cheminformatics, 2012 4:27.
+        http://dx.doi.org/10.1186/1758-2946-4-27
+
+        Parameters
+        ----------
+        molecule : oddt.toolkit.Molecule
+            Molecule to compute USRCAT shape descriptor
+
+        Returns
+        -------
+        shape_descriptor : numpy array, shape = (60)
+            Array describing shape of molecule
+    """
+    all_atoms_shape, points = common_usr(molecule)
+    ctd, cst, fct, ftf = points
+    hydrophobic_shape = common_usr(
+        molecule, ctd, cst, fct, ftf, 'ishydrophobe')[0]
```

Comment:
According to publication this should also contain halogens. Please add 'ishydrophobe' OR 'ishalogen' bool to mask.